Report portal exception is not handled properly which leads failure to create testng report

There are many instances where report portal expects some value or due to some validations report portal throws exception but then this is not handled properly which cause failure in testng report creation. It leads complete failure to build. To avoid such issue we had to remove report portal listener itself.

This issue is high priority and should be resolved.

The error logs can be found as below: 
```console
18:53:59 06:23:59.310 [main] INFO  c.m.a.cms.helper.APIHelperUtils - WRITING THE DATA INTO JSON FILE /tmp/testcaseresult.json
18:53:59 06:23:59.315 [RxComputationThreadPool-4] ERROR c.e.r.utils.SubscriptionUtils - [28] ReportPortal Finish test item execution error
18:53:59 com.epam.reportportal.exception.ReportPortalException: Report Portal returned error
18:53:59 Status code: 406
18:53:59 Status message: Not Acceptable
18:53:59 Error Message: Finish time 'Thu Jan 01 00:00:00 UTC 1970' is earlier than start time 'Tue Aug 13 13:15:24 UTC 2019' for resource with ID '5d52b8199bd1160001b8e043'
18:53:59 Error Type: FINISH_TIME_EARLIER_THAN_START_TIME
18:53:59 
18:53:59 	at com.epam.reportportal.service.ReportPortalErrorHandler.handleError(ReportPortalErrorHandler.java:83) ~[client-java-4.0.10.jar:na]
18:53:59 	at com.epam.reportportal.service.ReportPortalErrorHandler.handle(ReportPortalErrorHandler.java:59) ~[client-java-4.0.10.jar:na]
18:53:59 	at com.epam.reportportal.restendpoint.http.HttpClientRestEndpoint$1.subscribe(HttpClientRestEndpoint.java:517) ~[client-java-4.0.10.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeCreate.subscribeActual(MaybeCreate.java:45) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.Maybe.subscribe(Maybe.java:4096) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeCache.subscribeActual(MaybeCache.java:77) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.Maybe.subscribe(Maybe.java:4096) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26) ~[rxjava-2.1.16.jar:na]
18:53:59 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_144]
18:53:59 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_144]
18:53:59 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_144]
18:53:59 	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_144]
18:53:59 06:23:59.315 [RxComputationThreadPool-2] ERROR c.e.r.utils.SubscriptionUtils - [17] ReportPortal Finish test item execution error
18:53:59 com.epam.reportportal.exception.ReportPortalException: Report Portal returned error
18:53:59 Status code: 406
18:53:59 Status message: Not Acceptable
18:53:59 Error Message: Finish time 'Thu Jan 01 00:00:00 UTC 1970' is earlier than start time 'Tue Aug 13 13:21:31 UTC 2019' for resource with ID '5d52b9899bd1160001b8f454'
18:53:59 Error Type: FINISH_TIME_EARLIER_THAN_START_TIME
18:53:59 
18:53:59 	at com.epam.reportportal.service.ReportPortalErrorHandler.handleError(ReportPortalErrorHandler.java:83) ~[client-java-4.0.10.jar:na]
18:53:59 	at com.epam.reportportal.service.ReportPortalErrorHandler.handle(ReportPortalErrorHandler.java:59) ~[client-java-4.0.10.jar:na]
18:53:59 	at com.epam.reportportal.restendpoint.http.HttpClientRestEndpoint$1.subscribe(HttpClientRestEndpoint.java:517) ~[client-java-4.0.10.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeCreate.subscribeActual(MaybeCreate.java:45) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.Maybe.subscribe(Maybe.java:4096) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeCache.subscribeActual(MaybeCache.java:77) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.Maybe.subscribe(Maybe.java:4096) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38) ~[rxjava-2.1.16.jar:na]
18:53:59 	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26) ~[rxjava-2.1.16.jar:na]
18:53:59 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_144]
18:53:59 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_144]
18:53:59 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_144]
18:53:59 	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_144]
18:54:00 
18:54:00 Results :
18:54:00 
18:54:00 Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
18:54:00 
18:54:00 [INFO] ------------------------------------------------------------------------
18:54:00 [INFO] BUILD FAILURE
18:54:00 [INFO] ------------------------------------------------------------------------
18:54:00 [INFO] Total time: 01:24 h
18:54:00 [INFO] Finished at: 2019-08-13T06:23:59-07:00
18:54:00 [INFO] Final Memory: 83M/1008M
18:54:00 [INFO] ------------------------------------------------------------------------
18:54:00 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project cms-cmsfr: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: There was an error in the forked process
18:54:00 [ERROR] java.lang.IllegalArgumentException: ItemID should not be null
18:54:00 [ERROR] at rp.com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
18:54:00 [ERROR] at com.epam.reportportal.service.LaunchImpl.finishTestItem(LaunchImpl.java:247)
18:54:00 [ERROR] at com.epam.reportportal.testng.TestNGService.finishTestMethod(TestNGService.java:152)
18:54:00 [ERROR] at com.epam.reportportal.testng.BaseTestNGListener.onTestFailure(BaseTestNGListener.java:93)
18:54:00 [ERROR] at org.testng.internal.TestListenerHelper.runTestListeners(TestListenerHelper.java:66)
18:54:00 [ERROR] at org.testng.internal.TestInvoker.runTestResultListener(TestInvoker.java:210)
18:54:00 [ERROR] at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:814)
18:54:00 [ERROR] at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:145)
18:54:00 [ERROR] at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
18:54:00 [ERROR] at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
18:54:00 [ERROR] at java.util.ArrayList.forEach(ArrayList.java:1249)
18:54:00 [ERROR] at org.testng.TestRunner.privateRun(TestRunner.java:770)
18:54:00 [ERROR] at org.testng.TestRunner.run(TestRunner.java:591)
18:54:00 [ERROR] at org.testng.SuiteRunner.runTest(SuiteRunner.java:402)
18:54:00 [ERROR] at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:396)
18:54:00 [ERROR] at org.testng.SuiteRunner.privateRun(SuiteRunner.java:355)
18:54:00 [ERROR] at org.testng.SuiteRunner.run(SuiteRunner.java:304)
18:54:00 [ERROR] at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
18:54:00 [ERROR] at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
18:54:00 [ERROR] at org.testng.TestNG.runSuitesSequentially(TestNG.java:1180)
18:54:00 [ERROR] at org.testng.TestNG.runSuitesLocally(TestNG.java:1102)
18:54:00 [ERROR] at org.testng.TestNG.runSuites(TestNG.java:1032)
18:54:00 [ERROR] at org.testng.TestNG.run(TestNG.java:1000)
18:54:00 [ERROR] at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:281)
18:54:00 [ERROR] at org.apache.maven.surefire.testng.TestNGXmlTestSuite.execute(TestNGXmlTestSuite.java:75)
18:54:00 [ERROR] at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:121)
18:54:00 [ERROR] at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
18:54:00 [ERROR] at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
18:54:00 [ERROR] at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
18:54:00 [ERROR] -> [Help 1]
18:54:00 [ERROR] 
18:54:00 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
18:54:00 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
18:54:00 [ERROR] 
18:54:00 [ERROR] For more information about the errors and possible solutions, please read the following articles:
18:54:00 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
18:54:01 
18:54:01 [SSH] completed
18:54:01 [SSH] exit-status: 0
18:54:01 
18:54:01 [SSH] script:
```

--- a/src/main/java/com/epam/reportportal/service/LaunchImpl.java
+++ b/src/main/java/com/epam/reportportal/service/LaunchImpl.java
@@ -37,13 +37,12 @@ import io.reactivex.schedulers.Schedulers;
 import javax.validation.constraints.NotNull;
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.concurrent.*;
 
 import static com.epam.reportportal.service.LoggingCallback.*;
-import static com.epam.reportportal.utils.SubscriptionUtils.logCompletableResults;
-import static com.epam.reportportal.utils.SubscriptionUtils.logMaybeResults;
+import static com.epam.reportportal.utils.SubscriptionUtils.*;
 import static com.google.common.collect.Lists.newArrayList;
+import static java.util.Objects.requireNonNull;
 
 /**
  * @author Andrei Varabyeu
@@ -103,9 +102,9 @@ public class LaunchImpl extends Launch {
 	protected LaunchImpl(@NotNull final ReportPortalClient reportPortalClient, @NotNull final ListenerParameters parameters,
 			@NotNull final StartLaunchRQ rq, @NotNull final ExecutorService executorService) {
 		super(parameters);
-		rpClient = Objects.requireNonNull(reportPortalClient, "RestEndpoint shouldn't be NULL");
-		Objects.requireNonNull(parameters, "Parameters shouldn't be NULL");
-		executor = Objects.requireNonNull(executorService);
+		rpClient = requireNonNull(reportPortalClient, "RestEndpoint shouldn't be NULL");
+		requireNonNull(parameters, "Parameters shouldn't be NULL");
+		executor = requireNonNull(executorService);
 		scheduler = createScheduler(executor);
 
 		LOGGER.info("Rerun: {}", parameters.isRerun());
@@ -143,9 +142,9 @@ public class LaunchImpl extends Launch {
 	protected LaunchImpl(@NotNull final ReportPortalClient reportPortalClient, @NotNull final ListenerParameters parameters,
 			@NotNull final Maybe<String> launchMaybe, @NotNull final ExecutorService executorService) {
 		super(parameters);
-		rpClient = Objects.requireNonNull(reportPortalClient, "RestEndpoint shouldn't be NULL");
-		Objects.requireNonNull(parameters, "Parameters shouldn't be NULL");
-		executor = Objects.requireNonNull(executorService);
+		rpClient = requireNonNull(reportPortalClient, "RestEndpoint shouldn't be NULL");
+		requireNonNull(parameters, "Parameters shouldn't be NULL");
+		executor = requireNonNull(executorService);
 		scheduler = createScheduler(executor);
 
 		LOGGER.info("Rerun: {}", parameters.isRerun());
@@ -217,6 +216,11 @@ public class LaunchImpl extends Launch {
 		}
 	}
 
+	private static <T> Maybe<T> createErrorResponse(Throwable cause) {
+		LOGGER.error(cause.getMessage(), cause);
+		return createConstantMaybe(cause);
+	}
+
 	/**
 	 * Starts new test item in ReportPortal asynchronously (non-blocking)
 	 *
@@ -224,15 +228,19 @@ public class LaunchImpl extends Launch {
 	 * @return Test Item ID promise
 	 */
 	public Maybe<String> startTestItem(final StartTestItemRQ rq) {
+		if (rq == null) {
+			/*
+			 * This usually happens when we have a bug inside an agent or supported framework. But in any case we shouldn't rise an exception,
+			 * since we are reporting tool and our problems	should not fail launches.
+			 */
+			return createErrorResponse(new NullPointerException("StartTestItemRQ should not be null"));
+		}
 
-		final Maybe<String> testItem = launch.flatMap(new Function<String, Maybe<String>>() {
-			@Override
-			public Maybe<String> apply(String launchId) {
-				rq.setLaunchUuid(launchId);
-				return rpClient.startTestItem(rq).retry(DEFAULT_REQUEST_RETRY).doOnSuccess(logCreated("item")).map(TO_ID);
-
-			}
+		Maybe<String> testItem = launch.flatMap((Function<String, Maybe<String>>) launchId -> {
+			rq.setLaunchUuid(launchId);
+			return rpClient.startTestItem(rq).retry(DEFAULT_REQUEST_RETRY).doOnSuccess(logCreated("item")).map(TO_ID);
 		}).cache();
+
 		testItem.subscribeOn(scheduler).subscribe(logMaybeResults("Start test item"));
 		QUEUE.getUnchecked(testItem).addToQueue(testItem.ignoreElement().onErrorComplete());
 		LoggingContext.init(launch, testItem, rpClient, scheduler, getParameters().getBatchLogsSize(), getParameters().isConvertImage());
@@ -241,12 +249,7 @@ public class LaunchImpl extends Launch {
 	}
 
 	public Maybe<String> startTestItem(final Maybe<String> parentId, final Maybe<String> retryOf, final StartTestItemRQ rq) {
-		return retryOf.flatMap(new Function<String, Maybe<String>>() {
-			@Override
-			public Maybe<String> apply(String s) {
-				return startTestItem(parentId, rq);
-			}
-		}).cache();
+		return retryOf.flatMap((Function<String, Maybe<String>>) s -> startTestItem(parentId, rq)).cache();
 	}
 
 	/**
@@ -289,9 +292,17 @@ public class LaunchImpl extends Launch {
 	 * @param rq   Finish request
 	 * @return a Finish Item response promise
 	 */
-	public Maybe<OperationCompletionRS> finishTestItem(@NotNull final Maybe<String> item, @NotNull final FinishTestItemRQ rq) {
-		Objects.requireNonNull(item, "ItemID should not be null");
-		Objects.requireNonNull(rq, "FinishTestItemRQ should not be null");
+	public Maybe<OperationCompletionRS> finishTestItem(final Maybe<String> item, final FinishTestItemRQ rq) {
+		if (item == null) {
+			/*
+			 * This usually happens when we have a bug inside an agent or supported framework. But in any case we shouldn't rise an exception,
+			 * since we are reporting tool and our problems	should not fail launches.
+			 */
+			return createErrorResponse(new NullPointerException("ItemID should not be null"));
+		}
+		if (rq == null) {
+			return createErrorResponse(new NullPointerException("FinishTestItemRQ should not be null"));
+		}
 
 		if (ItemStatus.SKIPPED.name().equals(rq.getStatus()) && !getParameters().getSkippedAnIssue()) {
 			Issue issue = new Issue();
@@ -346,7 +357,7 @@ public class LaunchImpl extends Launch {
 	 */
 	protected static class TreeItem {
 		private volatile Maybe<String> parent;
-		private final List<Completable> children = new CopyOnWriteArrayList<Completable>();
+		private final List<Completable> children = new CopyOnWriteArrayList<>();
 
 		public LaunchImpl.TreeItem withParent(Maybe<String> parent) {
 			this.parent = parent;

--- a/src/main/java/com/epam/reportportal/utils/SubscriptionUtils.java
+++ b/src/main/java/com/epam/reportportal/utils/SubscriptionUtils.java
@@ -17,6 +17,7 @@ package com.epam.reportportal.utils;
 
 import io.reactivex.CompletableObserver;
 import io.reactivex.FlowableSubscriber;
+import io.reactivex.Maybe;
 import io.reactivex.MaybeObserver;
 import io.reactivex.disposables.Disposable;
 import org.reactivestreams.Subscription;
@@ -101,5 +102,19 @@ public class SubscriptionUtils {
 				LOGGER.debug("{} completed", type);
 			}
 		};
+	}
+
+	public static <T> Maybe<T> createConstantMaybe(final T rs) {
+		return Maybe.create(emitter -> {
+			emitter.onSuccess(rs);
+			emitter.onComplete();
+		});
+	}
+
+	public static <T> Maybe<T> createConstantMaybe(final Throwable exception) {
+		return Maybe.create(emitter -> {
+			emitter.onError(exception);
+			emitter.onComplete();
+		});
 	}
 }

--- a/src/test/java/com/epam/reportportal/service/LaunchNullCheckTest.java
+++ b/src/test/java/com/epam/reportportal/service/LaunchNullCheckTest.java
@@ -1,0 +1,99 @@
+/*
+ *  Copyright 2020 EPAM Systems
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.epam.reportportal.service;
+
+import com.epam.ta.reportportal.ws.model.OperationCompletionRS;
+import io.reactivex.Maybe;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import static com.epam.reportportal.test.TestUtils.*;
+
+public class LaunchNullCheckTest {
+
+	@Mock
+	private ReportPortalClient rpClient;
+
+	private final ExecutorService executor = Executors.newSingleThreadExecutor();
+
+	@BeforeEach
+	public void prepare() {
+		simulateStartLaunchResponse(rpClient);
+	}
+
+	@AfterEach
+	public void tearDown() throws InterruptedException {
+		executor.shutdown();
+		if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+			executor.shutdownNow();
+		}
+	}
+
+	// See https://github.com/reportportal/client-java/issues/99
+	@Test
+	public void launch_should_not_throw_any_exceptions_if_an_agent_bypasses_null_start_rq_on_start_item() {
+		Launch launch = new LaunchImpl(rpClient, STANDARD_PARAMETERS, standardLaunchRequest(STANDARD_PARAMETERS), executor);
+
+		launch.start();
+		Maybe<String> result = launch.startTestItem(null);
+
+		Assertions.assertThrows(NullPointerException.class, result::blockingGet);
+	}
+
+	@Test
+	public void launch_should_not_throw_any_exceptions_if_an_agent_bypasses_null_parent_item_id_on_start_item() {
+		Launch launch = new LaunchImpl(rpClient, STANDARD_PARAMETERS, standardLaunchRequest(STANDARD_PARAMETERS), executor);
+		simulateStartTestItemResponse(rpClient);
+
+		launch.start();
+		launch.startTestItem(standardStartSuiteRequest());
+		Maybe<String> result = launch.startTestItem(null, null);
+
+		Assertions.assertThrows(NullPointerException.class, result::blockingGet);
+	}
+
+	@Test
+	public void launch_should_not_throw_any_exceptions_if_an_agent_bypasses_null_item_id_on_finish_item() {
+		Launch launch = new LaunchImpl(rpClient, STANDARD_PARAMETERS, standardLaunchRequest(STANDARD_PARAMETERS), executor);
+		simulateStartTestItemResponse(rpClient);
+
+		launch.start();
+		launch.startTestItem(standardStartSuiteRequest());
+		Maybe<OperationCompletionRS> result = launch.finishTestItem(null, positiveFinishRequest());
+
+		Assertions.assertThrows(NullPointerException.class, result::blockingGet);
+	}
+
+	@Test
+	public void launch_should_not_throw_any_exceptions_if_an_agent_bypasses_null_finish_rq_on_finish_item() {
+		Launch launch = new LaunchImpl(rpClient, STANDARD_PARAMETERS, standardLaunchRequest(STANDARD_PARAMETERS), executor);
+		simulateStartTestItemResponse(rpClient);
+
+		launch.start();
+		launch.startTestItem(standardStartSuiteRequest());
+		Maybe<OperationCompletionRS> result = launch.finishTestItem(launch.startTestItem(standardStartSuiteRequest()), null);
+
+		Assertions.assertThrows(NullPointerException.class, result::blockingGet);
+	}
+}

--- a/src/test/java/com/epam/reportportal/service/LaunchTest.java
+++ b/src/test/java/com/epam/reportportal/service/LaunchTest.java
@@ -33,21 +33,31 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 
 import static com.epam.reportportal.test.TestUtils.*;
+import static com.epam.reportportal.utils.SubscriptionUtils.createConstantMaybe;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.*;
 
 public class LaunchTest {
 
-	private static final ErrorRS ERROR_RS;
+	private static final ErrorRS START_ERROR_RS;
+	private static final ErrorRS FINISH_ERROR_RS;
 
 	static {
-		ERROR_RS = new ErrorRS();
-		ERROR_RS.setErrorType(ErrorType.INCORRECT_REQUEST);
-		ERROR_RS.setMessage("Incorrect Request. [Value is not allowed for field 'status'.]");
+		START_ERROR_RS = new ErrorRS();
+		START_ERROR_RS.setErrorType(ErrorType.INCORRECT_REQUEST);
+		START_ERROR_RS.setMessage("Incorrect Request. [Value is not allowed for field 'status'.]");
+
+		FINISH_ERROR_RS = new ErrorRS();
+		FINISH_ERROR_RS.setErrorType(ErrorType.FINISH_TIME_EARLIER_THAN_START_TIME);
+		FINISH_ERROR_RS.setMessage(
+				"Finish time 'Thu Jan 01 00:00:00 UTC 1970' is earlier than start time 'Tue Aug 13 13:21:31 UTC 2019' for resource with ID '5d52b9899bd1160001b8f454'");
 	}
 
-	private static final ReportPortalException CLIENT_EXCEPTION = new ReportPortalException(400, "Bad Request", ERROR_RS);
+	private static final ReportPortalException START_CLIENT_EXCEPTION = new ReportPortalException(400, "Bad Request", START_ERROR_RS);
+
+	// taken from: https://github.com/reportportal/client-java/issues/99
+	private static final ReportPortalException FINISH_CLIENT_EXCEPTION = new ReportPortalException(406, "Not Acceptable", FINISH_ERROR_RS);
 
 	@Mock
 	private ReportPortalClient rpClient;
@@ -79,10 +89,10 @@ public class LaunchTest {
 		Maybe<String> testRs = launch.startTestItem(suiteRs, standardStartTestRequest());
 		Maybe<String> stepRs = launch.startTestItem(testRs, standardStartStepRequest());
 
-		when(rpClient.finishTestItem(eq(stepRs.blockingGet()), any())).thenThrow(CLIENT_EXCEPTION);
-		when(rpClient.finishTestItem(eq(testRs.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
-		when(rpClient.finishTestItem(eq(suiteRs.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
-		when(rpClient.finishLaunch(eq(launchUuid.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishTestItem(eq(stepRs.blockingGet()), any())).thenThrow(FINISH_CLIENT_EXCEPTION);
+		when(rpClient.finishTestItem(eq(testRs.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishTestItem(eq(suiteRs.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishLaunch(eq(launchUuid.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
 
 		launch.finishTestItem(stepRs, positiveFinishRequest());
 		launch.finishTestItem(testRs, positiveFinishRequest());
@@ -105,12 +115,12 @@ public class LaunchTest {
 		when(rpClient.startTestItem(eq(suiteRs.blockingGet()), any())).thenReturn(startTestItemResponse(UUID.randomUUID().toString()));
 		Maybe<String> testRs = launch.startTestItem(suiteRs, standardStartTestRequest());
 
-		when(rpClient.startTestItem(eq(testRs.blockingGet()), any())).thenThrow(CLIENT_EXCEPTION);
+		when(rpClient.startTestItem(eq(testRs.blockingGet()), any())).thenThrow(START_CLIENT_EXCEPTION);
 		Maybe<String> stepRs = launch.startTestItem(testRs, standardStartStepRequest());
 
-		when(rpClient.finishTestItem(eq(testRs.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
-		when(rpClient.finishTestItem(eq(suiteRs.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
-		when(rpClient.finishLaunch(eq(launchUuid.blockingGet()), any())).thenReturn(getConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishTestItem(eq(testRs.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishTestItem(eq(suiteRs.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
+		when(rpClient.finishLaunch(eq(launchUuid.blockingGet()), any())).thenReturn(createConstantMaybe(new OperationCompletionRS()));
 
 		launch.finishTestItem(stepRs, positiveFinishRequest());
 		launch.finishTestItem(testRs, positiveFinishRequest());

--- a/src/test/java/com/epam/reportportal/service/ReportPortalClientJoinTest.java
+++ b/src/test/java/com/epam/reportportal/service/ReportPortalClientJoinTest.java
@@ -21,6 +21,7 @@ import com.epam.reportportal.listeners.ItemStatus;
 import com.epam.reportportal.listeners.ListenerParameters;
 import com.epam.reportportal.service.launch.PrimaryLaunch;
 import com.epam.reportportal.service.launch.SecondaryLaunch;
+import com.epam.reportportal.utils.SubscriptionUtils;
 import com.epam.ta.reportportal.ws.model.ErrorRS;
 import com.epam.ta.reportportal.ws.model.FinishExecutionRQ;
 import com.epam.ta.reportportal.ws.model.OperationCompletionRS;
@@ -29,8 +30,6 @@ import com.epam.ta.reportportal.ws.model.item.ItemCreatedRS;
 import com.epam.ta.reportportal.ws.model.launch.LaunchResource;
 import com.epam.ta.reportportal.ws.model.launch.StartLaunchRQ;
 import io.reactivex.Maybe;
-import io.reactivex.MaybeEmitter;
-import io.reactivex.MaybeOnSubscribe;
 import io.reactivex.Scheduler;
 import io.reactivex.disposables.Disposable;
 import io.reactivex.functions.Consumer;
@@ -54,6 +53,7 @@ import java.util.concurrent.TimeUnit;
 
 import static com.epam.reportportal.test.TestUtils.simulateStartLaunchResponse;
 import static com.epam.reportportal.test.TestUtils.standardLaunchRequest;
+import static com.epam.reportportal.utils.SubscriptionUtils.createConstantMaybe;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.sameInstance;
@@ -86,7 +86,7 @@ public class ReportPortalClientJoinTest {
 	@AfterEach
 	public void tearDown() throws InterruptedException {
 		executor.shutdown();
-		if(!executor.awaitTermination(10, TimeUnit.SECONDS)) {
+		if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
 			executor.shutdownNow();
 		}
 	}
@@ -108,22 +108,12 @@ public class ReportPortalClientJoinTest {
 	private static Maybe<LaunchResource> getLaunchResponse(String id) {
 		final LaunchResource rs = new LaunchResource();
 		rs.setUuid(id);
-		return Maybe.create(new MaybeOnSubscribe<LaunchResource>() {
-			@Override
-			public void subscribe(final MaybeEmitter<LaunchResource> emitter) {
-				emitter.onSuccess(rs);
-				emitter.onComplete();
-			}
-		});
+		return createConstantMaybe(rs);
 	}
 
 	private static void simulateGetLaunchResponse(final ReportPortalClient client) {
-		when(client.getLaunchByUuid(anyString())).then(new Answer<Maybe<LaunchResource>>() {
-			@Override
-			public Maybe<LaunchResource> answer(InvocationOnMock invocation) {
-				return getLaunchResponse(invocation.getArgument(0).toString());
-			}
-		});
+		when(client.getLaunchByUuid(anyString())).then((Answer<Maybe<LaunchResource>>) invocation -> getLaunchResponse(invocation.getArgument(
+				0).toString()));
 	}
 
 	private static class StringConsumer implements Consumer<String> {
@@ -168,8 +158,8 @@ public class ReportPortalClientJoinTest {
 		return result;
 	}
 
-	private static List<Launch> createLaunchesNoGetLaunch(int num, ReportPortalClient rpClient, ListenerParameters params, LockFile lockFile,
-			ExecutorService executor) {
+	private static List<Launch> createLaunchesNoGetLaunch(int num, ReportPortalClient rpClient, ListenerParameters params,
+			LockFile lockFile, ExecutorService executor) {
 		simulateStartLaunchResponse(rpClient);
 		return createLaunchesNoStart(num, rpClient, params, lockFile, executor);
 	}
@@ -237,10 +227,7 @@ public class ReportPortalClientJoinTest {
 		launches.get(0).start();
 		launches.get(1).start();
 
-		when(rpClient.finishLaunch(
-				any(),
-				any(FinishExecutionRQ.class)
-		)).thenReturn(Maybe.create(e -> e.onSuccess(new OperationCompletionRS())));
+		when(rpClient.finishLaunch(any(), any(FinishExecutionRQ.class))).thenReturn(createConstantMaybe(new OperationCompletionRS()));
 
 		launches.get(0).finish(standardLaunchFinish());
 		launches.get(1).finish(standardLaunchFinish());
@@ -281,25 +268,14 @@ public class ReportPortalClientJoinTest {
 	}
 
 	private static void simulateStartItemResponse(final ReportPortalClient client, final String itemUuid) {
-		when(client.startTestItem(any(StartTestItemRQ.class))).then(new Answer<Maybe<ItemCreatedRS>>() {
-			@Override
-			public Maybe<ItemCreatedRS> answer(InvocationOnMock invocation) {
-				StartTestItemRQ rq = invocation.getArgument(0);
-				return standardItemResponse(itemUuid);
-			}
-		});
+		when(client.startTestItem(any(StartTestItemRQ.class))).then((Answer<Maybe<ItemCreatedRS>>) invocation -> standardItemResponse(
+				itemUuid));
 	}
 
 	private static Maybe<ItemCreatedRS> standardItemResponse(String id) {
-		final ItemCreatedRS rs = new ItemCreatedRS();
+		ItemCreatedRS rs = new ItemCreatedRS();
 		rs.setId(id);
-		return Maybe.create(new MaybeOnSubscribe<ItemCreatedRS>() {
-			@Override
-			public void subscribe(final MaybeEmitter<ItemCreatedRS> emitter) {
-				emitter.onSuccess(rs);
-				emitter.onComplete();
-			}
-		});
+		return createConstantMaybe(rs);
 	}
 
 	@Test
@@ -332,28 +308,14 @@ public class ReportPortalClientJoinTest {
 	}
 
 	private static Maybe<LaunchResource> getLaunchErrorResponse() {
-		return Maybe.create(new MaybeOnSubscribe<LaunchResource>() {
-			@Override
-			public void subscribe(final MaybeEmitter<LaunchResource> emitter) {
-				emitter.onError(new ReportPortalException(404, "Launch not found", new ErrorRS()));
-				emitter.onComplete();
-			}
-		});
+		return SubscriptionUtils.createConstantMaybe(new ReportPortalException(404, "Launch not found", new ErrorRS()));
 	}
 
 	private static void simulateGetLaunchByUuidResponse(ReportPortalClient client) {
-		Answer<Maybe<LaunchResource>> errorAnswer = new Answer<Maybe<LaunchResource>>() {
-			@Override
-			public Maybe<LaunchResource> answer(InvocationOnMock invocation) {
-				return getLaunchErrorResponse();
-			}
-		};
-		when(client.getLaunchByUuid(anyString())).then(errorAnswer).then(errorAnswer).then(new Answer<Maybe<LaunchResource>>() {
-			@Override
-			public Maybe<LaunchResource> answer(InvocationOnMock invocation) {
-				return getLaunchResponse(invocation.getArgument(0).toString());
-			}
-		});
+		Answer<Maybe<LaunchResource>> errorAnswer = invocation -> getLaunchErrorResponse();
+		when(client.getLaunchByUuid(anyString())).then(errorAnswer)
+				.then(errorAnswer)
+				.then((Answer<Maybe<LaunchResource>>) invocation -> getLaunchResponse(invocation.getArgument(0).toString()));
 	}
 
 	@Test
@@ -392,16 +354,16 @@ public class ReportPortalClientJoinTest {
 	public void test_two_launches_have_the_same_executors() {
 		List<Launch> launches = createLaunchesNoStart(2, rpClient, params, lockFile, executor);
 
-		assertThat(((LaunchImpl)launches.get(0)).getExecutor(), sameInstance(executor));
-		assertThat(((LaunchImpl)launches.get(1)).getExecutor(), sameInstance(executor));
+		assertThat(((LaunchImpl) launches.get(0)).getExecutor(), sameInstance(executor));
+		assertThat(((LaunchImpl) launches.get(1)).getExecutor(), sameInstance(executor));
 	}
 
 	@Test
 	public void test_two_launches_have_the_same_scheduler() {
 		List<Launch> launches = createLaunchesNoStart(2, rpClient, params, lockFile, executor);
 
-		Scheduler scheduler1 = ((LaunchImpl)launches.get(0)).getScheduler();
-		Scheduler scheduler2 = ((LaunchImpl)launches.get(1)).getScheduler();
+		Scheduler scheduler1 = ((LaunchImpl) launches.get(0)).getScheduler();
+		Scheduler scheduler2 = ((LaunchImpl) launches.get(1)).getScheduler();
 		assertThat(scheduler1, sameInstance(scheduler2));
 	}
 }

--- a/src/test/java/com/epam/reportportal/service/step/FileLocatorTest.java
+++ b/src/test/java/com/epam/reportportal/service/step/FileLocatorTest.java
@@ -41,6 +41,7 @@ import java.util.List;
 import java.util.UUID;
 import java.util.function.Supplier;
 
+import static com.epam.reportportal.utils.SubscriptionUtils.createConstantMaybe;
 import static java.util.Optional.ofNullable;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.*;
@@ -53,7 +54,7 @@ public class FileLocatorTest {
 	private final String testLaunchUuid = "launch" + UUID.randomUUID().toString().substring(6);
 	private final String testClassUuid = "class" + UUID.randomUUID().toString().substring(5);
 	private final String testMethodUuid = "test" + UUID.randomUUID().toString().substring(4);
-	private final Maybe<String> launchUuid = TestUtils.getConstantMaybe(testLaunchUuid);
+	private final Maybe<String> launchUuid = createConstantMaybe(testLaunchUuid);
 
 	@Mock
 	private ReportPortalClient rpClient;
@@ -62,14 +63,14 @@ public class FileLocatorTest {
 
 	private final Supplier<Maybe<ItemCreatedRS>> maybeSupplier = () -> {
 		String uuid = UUID.randomUUID().toString();
-		return TestUtils.getConstantMaybe(new ItemCreatedRS(uuid, uuid));
+		return createConstantMaybe(new ItemCreatedRS(uuid, uuid));
 	};
 
 	@BeforeEach
 	public void initMocks() {
-		Maybe<ItemCreatedRS> testMethodCreatedMaybe = TestUtils.getConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
+		Maybe<ItemCreatedRS> testMethodCreatedMaybe = createConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
 		when(rpClient.startTestItem(eq(testClassUuid), any())).thenReturn(testMethodCreatedMaybe);
-		when(rpClient.log(any(MultiPartRequest.class))).thenReturn(TestUtils.getConstantMaybe(new BatchSaveOperatingRS()));
+		when(rpClient.log(any(MultiPartRequest.class))).thenReturn(createConstantMaybe(new BatchSaveOperatingRS()));
 
 		// mock start nested steps
 		when(rpClient.startTestItem(eq(testMethodUuid),
@@ -78,7 +79,7 @@ public class FileLocatorTest {
 
 		ReportPortal rp = ReportPortal.create(rpClient, TestUtils.STANDARD_PARAMETERS);
 		Launch launch = rp.withLaunch(launchUuid);
-		launch.startTestItem(TestUtils.getConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
+		launch.startTestItem(createConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
 		sr = launch.getStepReporter();
 	}
 
@@ -90,8 +91,7 @@ public class FileLocatorTest {
 		verify(rpClient, timeout(1000).times(1)).log(logCaptor.capture());
 
 		MultiPartRequest logRq = logCaptor.getValue();
-		verifyFile(
-				logRq,
+		verifyFile(logRq,
 				Utils.readInputStreamToBytes(ofNullable(getClass().getClassLoader().getResourceAsStream("pug/lucky.jpg")).orElse(
 						EMPTY_STREAM)),
 				"lucky.jpg"
@@ -106,8 +106,7 @@ public class FileLocatorTest {
 		verify(rpClient, timeout(1000).times(1)).log(logCaptor.capture());
 
 		MultiPartRequest logRq = logCaptor.getValue();
-		verifyFile(
-				logRq,
+		verifyFile(logRq,
 				Utils.readInputStreamToBytes(ofNullable(getClass().getClassLoader().getResourceAsStream("pug/unlucky.jpg")).orElse(
 						EMPTY_STREAM)),
 				"unlucky.jpg"

--- a/src/test/java/com/epam/reportportal/service/step/ManualNestedStepTest.java
+++ b/src/test/java/com/epam/reportportal/service/step/ManualNestedStepTest.java
@@ -41,6 +41,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Supplier;
 
+import static com.epam.reportportal.utils.SubscriptionUtils.createConstantMaybe;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.equalTo;
@@ -53,7 +54,7 @@ public class ManualNestedStepTest {
 	private final String testLaunchUuid = "launch" + UUID.randomUUID().toString().substring(6);
 	private final String testClassUuid = "class" + UUID.randomUUID().toString().substring(5);
 	private final String testMethodUuid = "test" + UUID.randomUUID().toString().substring(4);
-	private final Maybe<String> launchUuid = TestUtils.getConstantMaybe(testLaunchUuid);
+	private final Maybe<String> launchUuid = createConstantMaybe(testLaunchUuid);
 
 	@Mock
 	private ReportPortalClient client;
@@ -66,14 +67,14 @@ public class ManualNestedStepTest {
 	private final List<Maybe<ItemCreatedRS>> createdStepsList = new ArrayList<>();
 	private final Supplier<Maybe<ItemCreatedRS>> maybeSupplier = () -> {
 		String uuid = UUID.randomUUID().toString();
-		Maybe<ItemCreatedRS> maybe = TestUtils.getConstantMaybe(new ItemCreatedRS(uuid, uuid));
+		Maybe<ItemCreatedRS> maybe = createConstantMaybe(new ItemCreatedRS(uuid, uuid));
 		createdStepsList.add(maybe);
 		return maybe;
 	};
 
 	@BeforeEach
 	public void initMocks() {
-		Maybe<ItemCreatedRS> testMethodCreatedMaybe = TestUtils.getConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
+		Maybe<ItemCreatedRS> testMethodCreatedMaybe = createConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
 		when(client.startTestItem(eq(testClassUuid), any())).thenReturn(testMethodCreatedMaybe);
 
 		// mock start nested steps
@@ -81,12 +82,12 @@ public class ManualNestedStepTest {
 		// mock finish nested steps
 		when(client.finishTestItem(any(String.class),
 				any(FinishTestItemRQ.class)
-		)).thenAnswer((Answer<Maybe<OperationCompletionRS>>) invocation -> TestUtils.getConstantMaybe(new OperationCompletionRS()));
+		)).thenAnswer((Answer<Maybe<OperationCompletionRS>>) invocation -> createConstantMaybe(new OperationCompletionRS()));
 
 		executor = Executors.newSingleThreadExecutor();
 		ReportPortal rp = ReportPortal.create(client, TestUtils.STANDARD_PARAMETERS, executor);
 		launch = rp.withLaunch(launchUuid);
-		testMethodUuidMaybe = launch.startTestItem(TestUtils.getConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
+		testMethodUuidMaybe = launch.startTestItem(createConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
 		sr = launch.getStepReporter();
 	}
 
@@ -156,8 +157,7 @@ public class ManualNestedStepTest {
 		ArgumentCaptor<FinishTestItemRQ> finishStepCaptor = ArgumentCaptor.forClass(FinishTestItemRQ.class);
 		verify(client, timeout(1000).times(1)).finishTestItem(eq(testMethodUuid), finishStepCaptor.capture());
 
-		assertThat(
-				"Parent test should fail if a nested step failed",
+		assertThat("Parent test should fail if a nested step failed",
 				finishStepCaptor.getValue().getStatus(),
 				equalTo(ItemStatus.FAILED.name())
 		);
@@ -172,7 +172,10 @@ public class ManualNestedStepTest {
 		sr.finishPreviousStep();
 
 		ArgumentCaptor<FinishTestItemRQ> finishStepCaptor = ArgumentCaptor.forClass(FinishTestItemRQ.class);
-		verify(client, timeout(1000).times(1)).finishTestItem(eq(createdStepsList.get(0).blockingGet().getUniqueId()), finishStepCaptor.capture());
+		verify(client, timeout(1000).times(1)).finishTestItem(
+				eq(createdStepsList.get(0).blockingGet().getUniqueId()),
+				finishStepCaptor.capture()
+		);
 
 		assertThat(finishStepCaptor.getValue().getStatus(), equalTo(ItemStatus.FAILED.name()));
 		assertThat("StepReporter should save parent failures", sr.isFailed(testMethodUuidMaybe), equalTo(Boolean.TRUE));

--- a/src/test/java/com/epam/reportportal/service/step/StepOrderTest.java
+++ b/src/test/java/com/epam/reportportal/service/step/StepOrderTest.java
@@ -37,6 +37,7 @@ import java.util.UUID;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.Supplier;
 
+import static com.epam.reportportal.utils.SubscriptionUtils.createConstantMaybe;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.*;
 import static org.mockito.Mockito.any;
@@ -47,7 +48,7 @@ public class StepOrderTest {
 	private final String testLaunchUuid = "launch" + UUID.randomUUID().toString().substring(6);
 	private final String testClassUuid = "class" + UUID.randomUUID().toString().substring(5);
 	private final String testMethodUuid = "test" + UUID.randomUUID().toString().substring(4);
-	private final Maybe<String> launchUuid = TestUtils.getConstantMaybe(testLaunchUuid);
+	private final Maybe<String> launchUuid = createConstantMaybe(testLaunchUuid);
 	private final AtomicInteger counter = new AtomicInteger();
 
 	@Mock
@@ -58,28 +59,30 @@ public class StepOrderTest {
 	private final List<Maybe<ItemCreatedRS>> createdStepsList = new ArrayList<>();
 	private final Supplier<Maybe<ItemCreatedRS>> maybeSupplier = () -> {
 		String uuid = UUID.randomUUID().toString();
-		Maybe<ItemCreatedRS> maybe = TestUtils.getConstantMaybe(new ItemCreatedRS(uuid, uuid));
+		Maybe<ItemCreatedRS> maybe = createConstantMaybe(new ItemCreatedRS(uuid, uuid));
 		createdStepsList.add(maybe);
 		return maybe;
 	};
 
 	@BeforeEach
 	public void initMocks() {
-		Maybe<ItemCreatedRS> testMethodCreatedMaybe = TestUtils.getConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
+		Maybe<ItemCreatedRS> testMethodCreatedMaybe = createConstantMaybe(new ItemCreatedRS(testMethodUuid, testMethodUuid));
 		when(client.startTestItem(eq(testClassUuid), any())).thenReturn(testMethodCreatedMaybe);
 
 		// mock start nested steps
-		when(client.startTestItem(eq(testMethodUuid),
+		when(client.startTestItem(
+				eq(testMethodUuid),
 				any()
 		)).thenAnswer((Answer<Maybe<ItemCreatedRS>>) invocation -> createdStepsList.get(counter.getAndIncrement()));
 		// mock finish nested steps
-		when(client.finishTestItem(any(String.class),
+		when(client.finishTestItem(
+				any(String.class),
 				any(FinishTestItemRQ.class)
-		)).thenAnswer((Answer<Maybe<OperationCompletionRS>>) invocation -> TestUtils.getConstantMaybe(new OperationCompletionRS()));
+		)).thenAnswer((Answer<Maybe<OperationCompletionRS>>) invocation -> createConstantMaybe(new OperationCompletionRS()));
 
 		ReportPortal rp = ReportPortal.create(client, TestUtils.STANDARD_PARAMETERS);
 		Launch launch = rp.withLaunch(launchUuid);
-		launch.startTestItem(TestUtils.getConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
+		launch.startTestItem(createConstantMaybe(testClassUuid), TestUtils.standardStartStepRequest());
 		sr = launch.getStepReporter();
 	}
 
@@ -102,7 +105,8 @@ public class StepOrderTest {
 		List<StartTestItemRQ> rqs = stepCaptor.getAllValues();
 		assertThat(rqs, hasSize(stepNum));
 		for (int i = 1; i < stepNum; i++) {
-			assertThat("Each nested step should not complete in the same millisecond, iteration: " + i,
+			assertThat(
+					"Each nested step should not complete in the same millisecond, iteration: " + i,
 					rqs.get(i - 1).getStartTime(),
 					not(equalTo(rqs.get(i).getStartTime()))
 			);

--- a/src/test/java/com/epam/reportportal/test/TestUtils.java
+++ b/src/test/java/com/epam/reportportal/test/TestUtils.java
@@ -19,6 +19,7 @@ package com.epam.reportportal.test;
 import com.epam.reportportal.listeners.ListenerParameters;
 import com.epam.reportportal.restendpoint.http.MultiPartRequest;
 import com.epam.reportportal.service.ReportPortalClient;
+import com.epam.reportportal.utils.SubscriptionUtils;
 import com.epam.ta.reportportal.ws.model.BatchElementCreatedRS;
 import com.epam.ta.reportportal.ws.model.BatchSaveOperatingRS;
 import com.epam.ta.reportportal.ws.model.FinishTestItemRQ;
@@ -64,14 +65,7 @@ public class TestUtils {
 	public static Maybe<StartLaunchRS> startLaunchResponse(String id) {
 		final StartLaunchRS rs = new StartLaunchRS();
 		rs.setId(id);
-		return getConstantMaybe(rs);
-	}
-
-	public static <T> Maybe<T> getConstantMaybe(final T rs) {
-		return Maybe.create(emitter -> {
-			emitter.onSuccess(rs);
-			emitter.onComplete();
-		});
+		return SubscriptionUtils.createConstantMaybe(rs);
 	}
 
 	public static void simulateStartLaunchResponse(final ReportPortalClient client) {
@@ -95,7 +89,7 @@ public class TestUtils {
 	public static Maybe<ItemCreatedRS> startTestItemResponse(String id) {
 		final ItemCreatedRS rs = new ItemCreatedRS();
 		rs.setId(id);
-		return getConstantMaybe(rs);
+		return SubscriptionUtils.createConstantMaybe(rs);
 	}
 
 	public static void simulateStartTestItemResponse(final ReportPortalClient client) {
@@ -157,7 +151,7 @@ public class TestUtils {
 	public static Maybe<BatchSaveOperatingRS> batchLogResponse(List<String> ids) {
 		final BatchSaveOperatingRS rs = new BatchSaveOperatingRS();
 		ids.forEach(i -> rs.addResponse(new BatchElementCreatedRS(i)));
-		return getConstantMaybe(rs);
+		return SubscriptionUtils.createConstantMaybe(rs);
 	}
 
 	@SuppressWarnings("unchecked")

